### PR TITLE
fix weird "pending" behavior; closes #2286

### LINKED ISSUE #2286: Regression after doesn't execute if tests skipped in beforeEach
```js

after(function () {
  console.log('after outside suite')
})

describe('suite', function () {
  var count = 0

  beforeEach(function () {
    if (count > 2) this.skip()
  })

  after(function () {
    console.log('after in suite')
  })

  it('test 1', function () { ++count })
  it('test 2', function () { ++count })
  it('test 3', function () { ++count })
  it('test 4', function () { ++count })
  it('test 5', function () { ++count })
  it('test 6', function () { ++count })
})
```
will not output the `after in suite` since v2.5.0 due to PR https://github.com/mochajs/mocha/pull/2081 (found via `git bisect`)

reported by @blueice349, confirmed by myself

### PR DESCRIPTION
**Pull request by boneskull**
continuation of #2571 with behavior suggested by @laughinghan.

this is targeted to `beta` branch for pre-release

### CODE REVIEW

**Comment by boneskull** [comment]
@ScottFreeCode any further comments?

File: test/integration/regression.spec.js [hunk 1]
```diff
@@ -54,17 +55,175 @@ describe('regressions', function () {
...
+       *   "afterAll" x 2.  There should be 10 items in this array.
+       * @param {string} [mode=always] - One of 'always' or 'once'
+       */
+      function testPendingRunnables (name, expected, mode) {
+        mode = mode || 'always';
+        var spies = [];
+
+        function spy (skip) {
+          function wrapped () {
+            if ((!wrapped.runCount++ || mode === 'always') && skip) {
```
Comment (fearphage) [inline test/integration/regression.spec.js]: `wrapped.runCount++ === 0` would add a bit more clarity and make it very explicit what you're looking for.

File: test/integration/regression.spec.js [hunk 2]
```diff
@@ -54,17 +55,175 @@ describe('regressions', function () {
...
+              return Boolean(spy.runCount);
+            })
+              .should
+              .deepEqual(expected);
+          });
+        });
+      }
+
+      testPendingRunnables('beforeAll', [
+        true, // beforeAll
```
Comment (fearphage) [inline test/integration/regression.spec.js]: Booleans that map to calls are not intuitive. What if you used strings instead?

Possibly use `beforeAll` and `beforeAll-called-skip` (or something similar) to denote the state?

Then you wouldn't need comments to explain what's everything means.

File: test/integration/regression.spec.js (hunk 1 above)
Comment (boneskull) [inline test/integration/regression.spec.js]: Thanks for your thoughts!  

I usually just prefer to use `!` when a falsy check is adequate.  It's a common style, so I'm not too concerned about it.

Though, some prefer explicit addition (`+`) over increment operator(s).  This is more obvious for readers of JavaScript who are, say, unfamiliar with the difference between `foo++` and `++foo` (or even `+foo`).

But, if it's opaque, I'll find another way to skin this particular cat.

File: test/integration/regression.spec.js (hunk 2 above)
Comment (boneskull) [inline test/integration/regression.spec.js]: That's a good suggestion.  

This entire test is a pretty rotten hack, which I am unhappy with.  If anyone can think of a better solution--and/or implement the requisite harness(es)--I'd be grateful!

**Comment by craigtaub** [comment]
@boneskull struggling to confirm if this is still useful. Could you confirm pls?

**Pull request closed without merging**